Extensions: Adding FileProviders and FileSystemGlobbing

maryamariyan opened the pull request:
- [x] Initial commits are bulk changes on remaining (upcoming projects) (First 5 commits)
- [x] FileSystemGlobbing ([6th commit](https://github.com/dotnet/runtime/pull/33212/commits/7140ac0f411ff595b6a753159f53d6afb3062121))
- [x] FileProviders [(7th commit](https://github.com/dotnet/runtime/pull/33212/commits/35d1037fcd5de14e9f3ee2ee534f44aace35a5a6))

Todo on this PR:
- [x] For FileSystemGlobbing: Fix packaging error + nit PR feedback changes ([8th commit](https://github.com/dotnet/runtime/pull/33212/commits/3560554a5f6cd1f9c6811894a0582e8af0ba1232))
- [x] For FileProviders.Physical project: Fix test project ([commit 11](https://github.com/dotnet/runtime/pull/33212/commits/ec06b61cb7062152718925f77c42b8abaf9869f2))
- [x] Bring over newer commit history from extensions repo
- [x] For FileProviders projects (Abstractions/Composite/Physical): add pkg ([9th commit](https://github.com/dotnet/runtime/pull/33212/commits/38a148c8a530212b2d055df3c472ae37c1ee4687))

For upcoming PRs:
- [x] Enable remaining packages: Configurations, Hosting, Logging, Options, DependencyInjection
there will be some tasks I'll get to once all projects are building, like:
- [ ] Adding solutions files
- [ ] Removing styling and whitespace nit errors skipped using NoWarns (refer to Directory.Build.props diff)
- [ ] Identifying and removing unused classes (especially in shared/extensions folders)

maryamariyan committed:
- [x] Bulk update all ref csproj except for FileProviders/FileSystemGlobbing

maryamariyan committed:
- Bulk move Descriptions to json file, all except for FileProviders and FileSystemGlobbing

maryamariyan committed:
- Bulk correct test project targetframeworks, all except for FileProviders and FileSystemGlobbing

maryamariyan committed:
- Bulk rename ref cs file and delete the extra duplicates
note: will be replaced by GenerateReferenceSource afterwards

maryamariyan committed:
- Bulk move/rename Resources file
from: Resources.resx to Resources\Strings.resx

maryamariyan committed:
- [x] Enables ref/src/test/pkg on FileSystemGlobbing
-- note: in a previous bulk commit, descriptions
for FileProviders and FileSystemGlobbing were already added in json file

- [ ] TODO: test packaging failed with message:
 Unable to find description for package Microsoft.Extensions.FileSystemGlobbing
 (perhaps because it wants to create package for netcoreapp)

maryamariyan committed:
Enables ref/src on FileProviders projects
- [x] Enable on Directory.Build.props
- [x] build ref/src projects and used GenerateReferenceSource
- [x] cleanup src csproj properties
- [x] set proper targetframeworks

TODO
- [ ] Add package, 3 projects go in one package
- [ ] Fix failing test on M.E.FP.Physical with message:
Could not load file or assembly 'Microsoft.Extensions.FileProviders.Abstractions, Version=5.0.0.0

ericstj commented:
> Fix runtime test error on FileProviders.Physical project

I'm not seeing this?

ericstj commented:
@steveharter @layomia looks like Json tests timed out here:
```
  Discovering: System.Text.Json.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Text.Json.Tests (found 1690 of 1718 test cases)
  Starting:    System.Text.Json.Tests (parallel test collections = on, max threads = 2)
   System.Text.Json.Tests: [Long Running Test] 'System.Text.Json.Serialization.Tests.StreamTests.RoundTripAsync', Elapsed: 00:03:45
   System.Text.Json.Tests: [Long Running Test] 'System.Text.Json.Serialization.Tests.StreamTests.RoundTripAsync', Elapsed: 00:05:45
   System.Text.Json.Tests: [Long Running Test] 'System.Text.Json.Serialization.Tests.StreamTests.RoundTripAsync', Elapsed: 00:07:45
   System.Text.Json.Tests: [Long Running Test] 'System.Text.Json.Serialization.Tests.StreamTests.RoundTripAsync', Elapsed: 00:09:45
   System.Text.Json.Tests: [Long Running Test] 'System.Text.Json.Serialization.Tests.StreamTests.RoundTripAsync', Elapsed: 00:11:45
   System.Text.Json.Tests: [Long Running Test] 'System.Text.Json.Serialization.Tests.StreamTests.RoundTripAsync', Elapsed: 00:13:45
```

ericstj commented on src/libraries/Directory.Build.props:
```diff
@@ -15,7 +17,7 @@
     <IsAspNetCoreApp>true</IsAspNetCoreApp>
     <!-- Temporarily ignore nit warning/errors (e.g. extra whitespace) from Extensions projects -->
-    <NoWarn Condition="!$(MSBuildProjectName.EndsWith('.Tests'))">$(NoWarn);SA1129;SA1028;SA1027;SA1121;CA1200</NoWarn>
+    <NoWarn Condition="!$(MSBuildProjectName.EndsWith('.Tests'))">$(NoWarn);SA1129;SA1028;SA1027;SA1121;CA1200;SA1000;CA1507;CA1802;CA1825</NoWarn>
```
For the curious (like me)
SA1129: DoNotUseDefaultValueTypeConstructor
SA1028: CodeMustNotContainTrailingWhitespace
SA1027: UseTabsCorrectly
SA1121: UseBuiltInTypeAlias
CA1200: Avoid using cref tags with a prefix
SA1000 : KeywordsMustBeSpacedCorrectly
CA1507: Use nameof in place of string
CA1802: Use Literals Where Appropriate
CA1825: Avoid zero-length array allocations

Most of these should be easy to come back and fix at once.

ericstj reviewed (commented):
sln files?

ericstj commented:
> Decide on how to do packaging that has 3 pairs of src/ref projects

Match what extensions did, which should be 1:1 library to package.

eerhardt commented on src/libraries/Microsoft.Extensions.Options.DataAnnotations/src/Microsoft.Extensions.Options.DataAnnotations.csproj:
```diff
@@ -1,13 +1,8 @@
     <TargetFrameworks>netstandard2.0;$(DefaultNetCoreTargetFramework)</TargetFrameworks>
     <TargetFrameworks Condition="'$(DotNetBuildFromSource)' == 'true'">$(DefaultNetCoreTargetFramework)</TargetFrameworks>
-    <GenerateDocumentationFile>true</GenerateDocumentationFile>
-    <PackageTags>aspnetcore;validation;options</PackageTags>
```
Where do these "PackageTags" go now? If we are removing them here, where should it go?

ericstj replied:
If we wanted to maintain them they'd go in the pkgproj, but we'd need to fix this: https://github.com/dotnet/arcade/blob/ed8c23ff7c4f4988a754ff9afc626571f16e3382/src/Microsoft.DotNet.Build.Tasks.Packaging/src/build/Packaging.targets#L52

ericstj replied:
Honestly I don't think they are worth preserving, but I'm happy to hear otherwise if folks feel strongly.

eerhardt replied:
@anurse @bradygaster - how much effort/thought went into the existing tags?

I agree we can easily dropp `aspnetcore` from these. But I'm not so sure about the other tags.

anurse replied:
Thought? Almost none ;). Feel free to change them at-will.

eerhardt commented on src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Properties/InternalsVisibleTo.cs:
```diff
@@ -0,0 +1,7 @@
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Microsoft.Extensions.FileSystemGlobbing.Tests, PublicKey=00240000048000009400000006020000002400005253413100040000010001004b86c4cb78549b34bab61a3b1800e23bfeb5b3ec390074041536a7e3cbd97f5f04cf0f857155a8928eaa29ebfd11cfbbad3ba70efea7bda3226c6a8d370a4cd303f714486b6ebc225985a638471e6ef571cc92a4613c00b8fa65d61ccee0cbe5f36330c9a01f4183559f1bef24cc2917c6d913e3a541333a1d05d9b ...
```
Why do you need both this and the 

```
 <InternalsVisibleTo Include="Microsoft.Extensions.FileSystemGlobbing.Tests" />
```

line in the .csproj?

eerhardt commented on src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Microsoft.Extensions.FileSystemGlobbing.csproj:
```diff
@@ -1,23 +1,19 @@
-    <AssemblyName>Microsoft.Extensions.FileSystemGlobbing</AssemblyName>
-    <Description>File system globbing to find files matching a specified pattern.</Description>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFrameworks>netstandard2.0</TargetFrameworks>
```
If there is only 1 TFM, do we have to use `TargetFrameworks`? There is a non-zero cost to using `TargetFrameworks` when you only have a single TFM.

ericstj replied:
Yes, we need to use TargetFramework in dotnet/runtime right now. https://github.com/dotnet/runtime/issues/31823

eerhardt replied:
That seems unfortunate. It has always been our public guidance to users to use `TargetFramework` when you only have a single TFM, and `TargetFrameworks` when you have multiple.

ericstj replied:
It's a known issue, we'll get around to fixing it.  For now, it's expected that in dotnet/runtime this will be the case.  /cc @Anipik @ViktorHofer

eerhardt commented on src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Microsoft.Extensions.FileSystemGlobbing.csproj:
```diff
@@ -1,23 +1,19 @@
 
   <ItemGroup>
-    <Compile Include="$(SharedSourceRoot)HashCodeCombiner\*.cs" />
+    <Compile Include="$(CommonPath)Extensions\HashCodeCombiner\HashCodeCombiner.cs">
```
We already have multiple "hashing helpers" code in the repo. Can we use one of the existing approaches instead of making yet another one?

[https://github.com/dotnet/runtime/blob/327801e8b838b97950cb8daaeef724acbe4e3c5f/src/libraries/System.Private.CoreLib/src/System/Numerics/Hashing/HashHelpers.cs#L7-L11](https://github.com/dotnet/runtime/blob/327801e8b838b97950cb8daaeef724acbe4e3c5f/src/libraries/System.Private.CoreLib/src/System/Numerics/Hashing/HashHelpers.cs#L7-L11)
[https://github.com/dotnet/runtime/blob/327801e8b838b97950cb8daaeef724acbe4e3c5f/src/libraries/System.Private.CoreLib/src/System/HashCode.cs#L51-L54](https://github.com/dotnet/runtime/blob/327801e8b838b97950cb8daaeef724acbe4e3c5f/src/libraries/System.Private.CoreLib/src/System/HashCode.cs#L51-L54)

(This doesn't need to be addressed in this PR, but potentially work for the future)

ericstj replied:
For any issues related to code that was moved over please file a follow up issue.  We don't want these PRs to be addressing code issues, only infra / build.

eerhardt replied:
https://github.com/dotnet/runtime/issues/33259

maryamariyan commented on src/libraries/pkg/descriptions.json:
```diff
@@ -14,6 +14,171 @@
+    "CommonTypes": []
+  },
+  {
+    "Name": "Microsoft.Extensions.FileSystemGlobbing.",
```
Awesome! thanks

maryamariyan committed:
For FileSystemGlobbing:
- Fix packaging issue
- Remove InternalsVisibleTo in src csproj
- Add sln file

maryamariyan commented:
> I'm not seeing this?

@ericstj that should show up in next PR

ericstj commented on src/libraries/Microsoft.Extensions.Configuration.Abstractions/ref/Microsoft.Extensions.Configuration.Abstractions.csproj:
```diff
@@ -1,14 +1,9 @@
-    <Compile Include="Microsoft.Extensions.Configuration.Abstractions.netcoreapp.cs" />
+  <ItemGroup>
+    <Compile Include="Microsoft.Extensions.Configuration.Abstractions.cs" />
     <Reference Include="Microsoft.Extensions.Primitives"  />
```
Ref projects need to use ProjectReference.

maryamariyan committed:
Add packaging for M.E.FileProviders.*:
- Abstractions
- Composite
- Physical

maryamariyan committed:
Switch Reference to ProjectReference in ref proj

steveharter commented:
> @steveharter @layomia looks like Json tests timed out here:
> System.Text.Json.Tests: [Long Running Test] 'System.Text.Json.Serialization.Tests.StreamTests.RoundTripAsync', Elapsed: 00:13:45

I have not seen a timeout here before.

I looked at the test and it doesn't seem possible to get a hang\deadlock.

The test uses ~.5MB -- is memory low on the hardware?

maryamariyan committed:
- Removed unnecessary helper files from src project

Done:
[x] bring over more history from extensions repo for
- TaskExtensions.cs
- ITestCondition.cs
- OperatingSystems.cs
- OSSkipConditionAttribute.c

maryamariyan committed:
Remove ParameterDefaultValue. As It was not needed.

ericstj commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/Microsoft.Extensions.FileProviders.Physical.Tests.csproj:
```diff
@@ -2,9 +2,35 @@
 
+  <ItemGroup>
+    <Compile Include="**\*.cs" />
+    <Compile Include="$(CommonPath)Extensions\Microsoft.AspNetCore.Testing\src\xunit\OSSkipConditionAttribute.cs">
```
I believe you might be missing TestingUtils in this path.  Can you fix that up and build these projects to make sure they're fixed before pushing?

maryamariyan committed:
Fix common TestingUtils path

ericstj commented:
This looks good now.  CI is going 📗 🚀

maryamariyan commented:
Looking over the failing tests on CI, e.g.: 

https://helix.dot.net/api/2019-06-17/jobs/013bd288-287d-4930-b113-d44f8ad18523/workitems/Microsoft.Extensions.FileProviders.Physical.Tests/console

maryamariyan committed:
Using PlatformSpecific instead of OSSkipCondition

maryamariyan commented:
The remaining failure on the CI is for System.IO.FileSystem.Watcher.Tests
https://github.com/dotnet/runtime/pull/33212/checks?check_run_id=504464776

This should be good to merge

ericstj commented:
Did you open an issue for the FileSystem.Watcher test failure?  Was there one already tracking it?

ericstj commented:
Here we go: https://github.com/dotnet/runtime/issues/24181
I've reopened it.

Please make sure you take some action when you observe a failure that isn't caused by your change.  That's how we get better 👍
